README
SchoolBell
==========

SchoolBell is a free and open source application to allow groups and
organizations to coordinate schedules and timetables.  Users may manage their
personal calendars, group calendars and calendars for resources, e.g., rooms,
projectors, etc, via a web interface, or using an iCalendar compliant client
such as Mozilla calendar or iCal.

SchoolBell is written in Python, and is a Zope 3 component as well as a
standalone calendaring server.  It is part of the SchoolTool project, funded
by the Shuttleworth Foundation.

Website: http://www.schooltool.org/schoolbell

Mailing list: http://lists.schooltool.org/mailman/listinfo/schooltool

Bug tracker: http://issues.schooltool.org/


System requirements
-------------------

- Python 2.3 or 2.4 (http://www.python.org/)
  (Debian users will need either python2.3 and python2.3-xml, or
  python2.4 and python2.4-xml packages)

- Zope X3 (http://www.zope.org/)
  (Users will require Zope X3 development trunk revision,
  Zope libraries of this revision are included with the distribution)

- libxml2 Python bindings (http://xmlsoft.org/).  Windows users can find
  binaries here:  http://users.skynet.be/sbi/libxml-python/

- (optional) the ReportLab Toolkit (http://www.reportlab.org), and
  Arial and Times New Roman TrueType fonts.  ReportLab is only needed if
  you want to generate PDF calendars.  To enable PDF support you will
  need to specify the path to fonts in the configuration file.

- Development environment and a C compiler
  (pre-compiled binaries are available for Windows users at the SchoolBell
  website)


Building SchoolBell
-------------------

Run 'make build update-translations' to build the necessary extension modules
and translations.

It is a good idea to run 'make test' and 'make ftest' to check if all the
essential unit and functional tests pass.


Installing and Configuring with Distutils (optional)
----------------------------------------------------

It is possible to install SchoolBell using python distutils.  To do this, you
have to use one of the distributed tarballs/zipfiles, which you unpack in a
directory with the tools appropriate to your OS.  First, it is necessary to
install the version of Zope included with the tarballs, this can be done by
running

python setup.py install

from the Zope3/ directory included with the SchoolBell distribution.  From the
top level directory of an unpacked SchoolBell tarball/zipfile SchoolBell can be
installed with:

python setup.py install

Finally, to setup a SchoolBell instance, copy the installed script into a
separate directory and create a config file with the name of the script and an
extension .conf.  A good template can be found in the top level of the tarball
(.conf.in). Running this script will then run an instance in that directory.

Those interested in installing into a non-standard location should investigate
the --paths and --default-config option for the install distutils commands.

NOTE: If you are using a Linux distribution for which a SchoolBell
package exists, you should consider using those packages instead.


Running SchoolBell
------------------

The top-level project directory contains the following executable Python
scripts:

  schoolbell-server.py      starts the SchoolBell server

The SchoolBell server automatically creates an empty database if it cannot find
an existing one.  You can customize the location of the database and a few
other parameters in a configuration file called schoolbell.conf.  There's
an example file called schoolbell.conf.in, you can simply rename it and modify
to suit your needs.

By default a user with manager privileges is created in the new database.
The username is 'manager', and the password is 'schoolbell'.

The default web application port is 7080.  Once the server is running, you can
connect to it with a web browser.

The Makefile also contains several shortcuts, which are kept mainly for
backwards-compatibility:

  make run                  starts the SchoolBell server
  make build                build SchoolBell and the included Zope 3 libraries
  make test                 run the unit test suite on SchoolBell
  make ftest                run the functional test suite on SchoolBell

These commands are superseded by executable Python scripts in the top-level
directory.


Project structure
-----------------

  GPL                   the GNU General Public License, version 2
  README                this file
  RELEASE               release notes for the latest release

  Makefile              makefile for building extension modules
  setup.py              distutils setup script for building extension modules
  test.py               test runner
  remove-stale-bytecode.py
                        script to remove stale *.pyc files

  schoolbell-server.py  script to start the SchoolBell server
  schoolbell.conf.in    sample configuration file for SchoolBell

  build/                temporary files are placed here during build process
  debian/               Debian packaging support
  doc/                  documentation
  src/                  source code
    schoolbell/         Python 'schoolbell' package
      app/              The SchoolBell application
      calendar/         SchoolBell calendaring library for Zope 3
      relationship/     The SchoolBell relationship library.


Testing
-------

There are two sets of automated tests: unit tests and functional tests.
Unit tests (sometimes also called programmer tests) test individual components
of the system.  Functional tests (also called customer or acceptance tests)
test only externally visible behaviour of the whole system.

Tests themselves are scattered through the whole source tree.  Subdirectories
named 'tests' contain unit tests, while subdirectories named 'ftests' contain
functional tests.

To run all unit tests, do

  python test.py -pv

To run all functional tests, do

  python test.py -fpv

The test runner has more options and features.  To find out about them, do

  python test.py -h

Functional tests are are not completely isolated.  Some functional tests
create named database state snapshots, while other tests reuse those snapshots
of known database state.  The test runner does not know which tests create
which snapshots, so if you want to run just one (or a couple) of functional
tests in isolation, it is likely that you will have first run the full suite
to create the necessary snapshots.  Do not restart the test server, or all
saved snapshots will be gone.


Virtual hosting
---------------

SchoolBell provides support for virtual hosting with Apache's
mod_rewrite and libproxy.  For example, let's say you have two
SchoolBell instances running on ports 7001 and 7002, and you want to
make them available as school1.example.org and school2.example.org, both
on port 80.  In order to do that, add the following to your Apache
configuration file:

  NameVirtualHost *:80

  <VirtualHost *:80>
    ServerName school1.example.org
    RewriteEngine On
    RewriteRule ^/(.*) http://localhost:7001/++vh++http:school1.example.org:80/++/$1 [P]
  </VirtualHost>

  <VirtualHost *:80>
    ServerName school2.example.org
    RewriteEngine On
    RewriteRule ^/(.*) http://localhost:7002/++vh++http:school2.example.org:80/++/$1 [P]
  </VirtualHost>

Also, enable mod_proxy and mod_rewrite.


You can also get SSL support in the same way.

  NameVirtualHost *:443

  <VirtualHost *:443>
    ServerName school1.example.org
    SSLEnable          # Apache 1.3
    # SSLEngine On     # Apache 2.0
    RewriteEngine On
    RewriteRule ^/(.*) http://localhost:7001/++vh++https:school1.example.org:443/++/$1 [P]
  </VirtualHost>

The web application interface also supports virtual hosting in this manner,
the only differing thing would be the local port number.


Copyright information
---------------------

SchoolBell is copyright (c) 2003--2005 Shuttleworth Foundation

All files in the src/schoolbell directory are part of SchoolBell, and are (c)
Shuttleworth Foundation.

Unless otherwise stated, files in src/schoolbell are released under the
terms of the GNU General Public License as published by the Free
Software Foundation; either version 2 of the License, or (at your option)
any later version.

Files in the following directories are (c) their respective owners. See
the individual files and directories for details of the licences.

  Zope3

Files in the same directory as this README file are (c) Shuttleworth
Foundation, with the exception of GPL, which is a copy of the Free Software
Foundation's General Public License, and is (c) FSF.

Files in the debian/ subdirectory are (c) Shuttleworth Foundation and
contributors.


SchoolBell is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

You can find the full copy of the GNU General Public License in a file called
GPL in the project directory.


CHANGES.txt
=======
CHANGES
=======

2.8.1 (unreleased)
------------------

- Nothing changed yet.


2.8.0 (2014-05-08)
------------------

- Added 'Added in error' state for temporal relationships
- Added What is this info to membership state edit view (https://launchpad.net/bugs/1308944)


2.6.4 (2014-05-01)
------------------

- Support CKEditor 4.x (https://launchpad.net/bugs/1309133)
- Delete schedules that reference a deleted timetable (https://launchpad.net/bugs/1280528)
- Improve tab style to make active tab more clear (https://launchpad.net/bugs/1023378)
- Timetables copied from active year retain old start/end dates (https://launchpad.net/bugs/1005962)
- Updated translations


2.7.0 (2014-04-12)
------------------

New features
++++++++++++

- Temporal relationships
- Checkboxes in relationship views
- Levels for courses and persons
- Parent access
- Background import
- Added descriptions for demographic fields


2.6.3 (2014-04-12)
------------------

New features
++++++++++++

- Switch to CKEditor (https://launchpad.net/bugs/485898)
- Celery 3 support (https://launchpad.net/bugs/1276384)
- Latest translations from Launchpad

Bugfixes
++++++++

- Fixed exception dates format in calendar widget of recurring events (https://launchpad.net/bugs/372889)
- Fixed export of deleted timetables (https://launchpad.net/bugs/1281335)
- Updated copyright (https://launchpad.net/bugs/1289277)
- Fixed term/section navigation in a different school year (https://launchpad.net/bugs/1281050)


2.6.2 (2014-02-15)
------------------

- Fix upgrade from a fresh install of version 2.6.0 (https://launchpad.net/bugs/1258529)
- Fix delete year/term if it contains non-ascii chars.
- Updated translations


2.6.1 (2013-11-21)
------------------

- Synchronize timezone set in preferences with timetables and schedules (https://launchpad.net/bugs/1239468)
- Set initial timezone from ``/etc/timezone`` (https://launchpad.net/bugs/1030211)
- Add SchoolTool versions to the Server page.
- Person cannot see his own preferences (https://launchpad.net/bugs/1247825)
- Do not escape extra spaces in comments (https://launchpad.net/bugs/1241022)
- Fix error when trying to replace existing school logo (https://launchpad.net/bugs/1250733)
- Restore the warning about unsaved changes in gradebook, broken after adding
  autocompletion (https://launchpad.net/bugs/1253070)
- Updated translations


2.6.0 (2013-10-10)
------------------

- Don't reverse name order (https://launchpad.net/bugs/1201887)
- Add "Name sorting" option to Server settings, display the selected name
  column first in person tables.
- Person search does not find by full name (https://launchpad.net/bugs/894798)
- Change manager's full name to "Default Manager" (https://launchpad.net/bugs/1227220)
- Fix timetable wizard for weekly timetables with different periods order.
- Fix crash when displaying timetables with only one day (https://launchpad.net/bugs/1235150)
- Fix navigation inside the School tab if a non-active school year is selected,
  or school years title contains accented characters (https://launchpad.net/bugs/1234690)
- Setting multiple languages in schooltool.conf breaks remote reports (https://launchpad.net/bugs/1229013)
- Improvements to PDF grid and table style.
- Extra info in traceback page and submit to launchpad button
- Updated translations, removed unsupported ones (less than 25% translated).


2.5.3 (2013-09-17)
------------------

Bugfixes
++++++++

- Fix integer demographic fields handling in importer (https://launchpad.net/bugs/1211613)
- Add Activate school year link in School tab under Customize (https://launchpad.net/bugs/1216992 )
- Fix links and Done buttons inside School tab when a different school year is selected
- Add manager to Site Managers and Clerks groups (https://launchpad.net/bugs/1216915)
- Access setting: allow instructor to add students to their section (https://launchpad.net/bugs/637759)
- Access setting: allow instructor to modify their section (https://launchpad.net/bugs/683313)
- Access setting: instructors can schedule their own sections (https://launchpad.net/bugs/604643)
- Display contact information of a person in profile.
- Updated translations

Internals
+++++++++

- Add ability to disable views in customized packages (https://launchpad.net/bugs/716346)


2.5.2 (2013-08-22)
------------------

- Autocompletion in gradebook for discrete score systems.
- TAB key in gradebook moves to the next cell.
- Further updates to permissions.
- Clerks and administrators now have appropriate access to parts of the School.
- Update GPL license and file headers from gnu.org (https://launchpad.net/bugs/1211145)
- Fix failures when creating timetables with rotating periods and weekly time slots (https://launchpad.net/bugs/1214131)
- Updated translations


2.5.1 (2013-08-01)
------------------

- Differentiate permissions of administrators, clerks, and managers.
- Scripts and styles for the new attendance features.
- Updated translations


2.5.0 (2013-06-03)
------------------

Reports
+++++++

- Background XLS import (disabled)
- Background XLS export
- Background report generation
- Multiple reports in a .zip archive
- Messages about finished background tasks
- Added resource library for filtered/chart reports
- Removed Person XML reports

Server
++++++

- There are now 4 processes, controlled by ``supervisor``: ``schooltool``, ``zeo``,
  ``celery_report``, and ``redis``.
- New ``make`` targets: start, stop, restart, rerun, status.
- You have to install ``redis-server`` package on the system. Other new
  dependencies are Python packages.
- Instances have to be updated to run this version. If you run from a sandbox,
  do ``bin/make-schooltool-instance instance``.


2.4.1 (2013-05-30)
------------------

Reports
+++++++

- Fixed report generation to not do network requests (https://launchpad.net/bugs/1185418).
- Fixed ratio of school logo.
- Fixed rendering of "&" in headers.

Bugfixes
++++++++


i18nextract.py
#
# SchoolTool - common information systems platform for school administration
# Copyright (c) 2005 Shuttleworth Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
"""
Customization of Zope's message string extraction module for schooltool and
schoolbell.

This is just a very simple script to be run from within an un-packed and built
schooltool checkout.

$Id$
"""
import os
import sys
import tokenize
import optparse

from zope.i18nmessageid.message import MessageFactory
from zope.app.locales.pygettext import make_escapes
from zope.app.locales.extract import find_files

_import_chickens = {}, {}, ("*",) # dead chickens needed by __import__

here = os.path.abspath(os.path.dirname(__file__))

# Remove this directory from path:
# (prevents conflicts between python's test module and the test runner)
sys.path[:] = [p for p in sys.path if os.path.abspath(p) != here]
# Lets add the Zope and schooltool paths to sys.path
zope3 = os.path.join(here, 'Zope3', 'src')
sys.path.insert(0, zope3)
st = os.path.join(here, 'src')
sys.path.insert(0, st)

from zope.app.locales import extract

# Monkey patch the Zope3 translation extraction machinery
extract.pot_header = """\
# SchoolTool - common information systems platform for school administration
# Copyright (c) 2005    Shuttleworth Foundation
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
msgid ""
msgstr ""
"Project-Id-Version: %(version)s\\n"
"POT-Creation-Date: %(time)s\\n"
"PO-Revision-Date: YEAR-MO-DA HO:MI+ZONE\\n"
"Last-Translator: FULL NAME <EMAIL@ADDRESS>\\n"
"Language-Team: Schooltool Development Team <schooltool-dev@schooltool.org>\\n"
"MIME-Version: 1.0\\n"
"Content-Type: text/plain; charset=%(charset)s\\n"
"Content-Transfer-Encoding: %(encoding)s\\n"
"Generated-By: i18nextract.py\\n"

"""

def get_version():
    version_file = os.path.join('src', 'schooltool', 'common', 'version.txt')
    f = open(version_file, 'r')
    result = f.read()
    f.close()
    return result

class POTMaker(extract.POTMaker):

    def _getProductVersion(self):
        return "SchoolTool Version %s" % get_version()

def py_strings(dir, domain="zope", exclude=(), verify_domain=False):
    """Retrieve all Python messages from `dir` that are in the `domain`.
    """
    eater = extract.TokenEater()
    make_escapes(0)
    for filename in find_files(
            dir, '*.py', exclude=('extract.py', 'pygettext.py')+tuple(exclude)):

        if verify_domain:
            common_path_lengths = [
                len(os.path.commonprefix([os.path.abspath(filename), path]))
                for path in sys.path]
            l = sorted(common_path_lengths)[-1]
            # remove .py ending from filenames
            # replace all path separators with a dot
            # remove the __init__ from the import path
            import_name = filename[l+1:-3].replace(os.path.sep, ".").replace(".__init__", "")

            try:
                module = __import__(import_name, *_import_chickens)
            except ImportError, e:
                # XXX if we can't import it - we assume that the domain is
                # the right one
                print >> sys.stderr, "Could not import %s" % import_name
            else:
                mf = getattr(module, '_', None)
                # XXX if _ is has no _domain set we assume that the domain
                # is the right one, so if you are using something non
                # MessageFactory you should set it's _domain attribute.
                if hasattr(mf, '_domain'):
                    if mf._domain != domain:
                        # domain mismatch - skip
                        continue

        fp = open(filename)

        try:
            eater.set_filename(filename)
            try:
                tokenize.tokenize(fp.readline, eater)
            except tokenize.TokenError, e:
                print >> sys.stderr, '%s: %s, line %d, column %d' % (
                    e[0], filename, e[1][0], e[1][1])
        finally:
            fp.close()
    return eater.getCatalog()

def write_pot(output_dir, path, domain, base_dir, site_zcml):
    # Create the POT
    output_file = os.path.join(path, output_dir, domain + '.pot')
    maker = POTMaker(output_file, path)
    maker.add(py_strings(path, domain, verify_domain=True), base_dir)
    maker.add(extract.zcml_strings(path, domain, site_zcml=site_zcml), base_dir)
    maker.add(extract.tal_strings(path, domain), base_dir)
    maker.write()

def parse_args(argv):
    """Parse the command line arguments"""
    parser = optparse.OptionParser(usage="usage: %prog [options]")
    # XXX ! separate these
    parser.add_option("--domain", dest="domain", default=None,
                      help="The domain that should be extracted")
    parser.add_option("--zcml", dest="zcml", default=None,
                      help="ZCML file to start the extraction")
    parser.add_option("--locales", dest="locales", default=None,
                      help="Locales directory to store the pot file")
    options, args = parser.parse_args(argv)
    assert len(args) == 1
    assert options.domain is not None
    assert options.zcml is not None
    assert options.locales is not None
    return options

if __name__ == '__main__':
    here = os.path.abspath(os.path.dirname(__file__))
    path = os.path.join(here, 'src')
    # get the command line arguments
    options = parse_args(sys.argv)
    base_dir = here # Comments are relative to the source checkouts so we are
                    # sure we don't have any absolute paths in there.
    output_dir = os.path.abspath(options.locales)
    write_pot(output_dir, path, options.domain, base_dir, options.zcml)
    print 'Extracted %s.pot to %s' % (options.domain, output_dir)
